### ISSUE TITLE
fails with 'zram: Cannot change disksize for initialized device'

### ISSUE THREAD

**Issue by cmurf** [systemd/zram-generator#7]
This works on Fedora 31, but fails on Fedora 32. 
zram-generator-0.1.2-1.module_f31+6778+c4fd1ff5.x86_64
zram-generator-0.1.2-1.fc32.x86_64

It looks like /dev/zram0 already has a size at the time the generator tries to set the disksize:

    Process: 725 ExecStart=/usr/bin/sh -c echo 8210350080 >/sys/block/zram0/disksize (code=exited, status=1/FAILURE)

And sure enough, at the command line, this fails every time until the zram device is reset. It can only be specified once. That suggests this /dev/zram0 already had a size by the time the generator tried to set it. Either the generator tried to set size twice (?) or maybe there's been a kernel change? Or race?

```
# journalctl -b -o short-monotonic | grep zram
[    8.434595] fmac.local kernel: zram: Added device: zram0
[    8.741960] fmac.local systemd-modules-load[613]: Inserted module 'zram'
[    9.132182] fmac.local kernel: zram0: detected capacity change from 0 to 8210350080
[    9.910055] fmac.local audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    9.910344] fmac.local audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    9.911191] fmac.local systemd[1]: Found device /dev/zram0.
[    9.912333] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[    9.921916] fmac.local systemd[1]: swap-create@zram0.service: Succeeded.
[    9.922243] fmac.local systemd[1]: Started Create swap on /dev/zram0.
[    9.925440] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[    9.956339] fmac.local kernel: zram: Cannot change disksize for initialized device
[    9.954646] fmac.local systemd[1]: swap-create@zram0.service: Main process exited, code=exited, status=1/FAILURE
[    9.954997] fmac.local systemd[1]: swap-create@zram0.service: Failed with result 'exit-code'.
[    9.958255] fmac.local systemd[1]: Failed to start Create swap on /dev/zram0.
[    9.998253] fmac.local audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[    9.998554] fmac.local systemd[1]: Dependency failed for Compressed swap on /dev/zram0.
[    9.998955] fmac.local systemd[1]: dev-zram0.swap: Job dev-zram0.swap/start failed with result 'dependency'.
[   10.000053] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[   10.035048] fmac.local kernel: zram: Cannot change disksize for initialized device
[   10.034499] fmac.local systemd[1]: swap-create@zram0.service: Main process exited, code=exited, status=1/FAILURE
[   10.035467] fmac.local systemd[1]: swap-create@zram0.service: Failed with result 'exit-code'.
[   10.037191] fmac.local systemd[1]: Failed to start Create swap on /dev/zram0.
[   10.040332] fmac.local systemd[1]: Dependency failed for Compressed swap on /dev/zram0.
[   10.044883] fmac.local systemd[1]: dev-zram0.swap: Job dev-zram0.swap/start failed with result 'dependency'.
```

**Comment by cmurf** [comment]
OK there's a race somewhere I think. Three boots, three different results:
a. Failure per above
b. Works as expected
c. Is setup and activate, but also failure message:

```
[chris@fmac ~]$ sudo journalctl -b -o short-monotonic | grep swap
[sudo] password for chris: 
[    2.707152] fmac.local kernel: zswap: loaded using pool lzo/zbud
[    8.113817] fmac.local systemd[1]: Created slice system-swap\x2dcreate.slice.
[    8.226654] fmac.local systemd[1]: Activating swap Swap Partition...
[    8.249986] fmac.local kernel: Adding 8391840k swap on /dev/sda6.  Priority:-2 extents:1 across:8391840k SSFS
[    8.377060] fmac.local systemd[1]: Activated swap Swap Partition.
[    9.870205] fmac.local kernel: Adding 8017916k swap on /dev/zram0.  Priority:-3 extents:1 across:8017916k SSFS
[   12.571435] fmac.local audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   12.571504] fmac.local audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   12.571577] fmac.local audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=swap-create@zram0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[   12.576372] fmac.local mkswap[656]: Setting up swapspace version 1, size = 7.7 GiB (8210345984 bytes)
[   12.576372] fmac.local mkswap[656]: no label, UUID=9cdcb8b1-b515-45e1-a8f1-d188e1e9ce5f
[   12.577886] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[   12.581690] fmac.local systemd[1]: swap-create@zram0.service: Succeeded.
[   12.581755] fmac.local systemd[1]: Started Create swap on /dev/zram0.
[   12.581795] fmac.local systemd[1]: Activating swap Compressed swap on /dev/zram0...
[   12.581858] fmac.local systemd[1]: Activated swap Compressed swap on /dev/zram0.
[   12.582208] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[   12.582532] fmac.local systemd[1]: swap-create@zram0.service: Main process exited, code=exited, status=1/FAILURE
[   12.582580] fmac.local systemd[1]: swap-create@zram0.service: Failed with result 'exit-code'.
[   12.582629] fmac.local systemd[1]: Failed to start Create swap on /dev/zram0.
``` 

What's with the audit showing service start stop start?

**Comment by cmurf** [comment]
Problem happens only on this one baremetal F32 system, but doesn't happen if I set systemd.log_level=debug which is suggestive of a race (the heavier logging seems to slow things down a lot and they tend to fall into place more consistently and as intended, is my speculation).

The problem doesn't happen on F31 baremetal systems, nor on F31 or F32 virtual machines.

**Comment by cmurf** [comment]
I've removed anaconda and zram packages; therefore their swap+zram implementations can't conflict.

It seems clear that swap-create@zram0.service is run twice. But I don't know why.

```
[    9.197460] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[   10.162519] fmac.local systemd[1]: swap-create@zram0.service: Succeeded.
[   10.162851] fmac.local systemd[1]: Started Create swap on /dev/zram0.
```
Later it starts again and fails. The second start is not expected, the failure is understandable because the zram device is already setup and can't be sized again.
```
[   10.172608] fmac.local systemd[1]: Starting Create swap on /dev/zram0...
[   10.206493] fmac.local systemd[1]: swap-create@zram0.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by nabijaczleweli** [comment]
The ideal solution IMO would be to reset the zram device after `dev-zramN.swap` finishes, but (and this may be because I'm not proficient enough at systemd) a way to do that didn't really occur to me, however resetting it *beforehand* also works.

The problem with doing it via `zramctl` is that `zramctl --reset` also removes the device node from `/dev`, and there's no way to recreate a *specific* device (see sexion 7 "Add/remove zram devices" [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/admin-guide/blockdev/zram.rst?id=5871023c3a316e8fe8c03e7bad1d35057e69ab40)).

The sysfs `/sys/block/zramN/reset` node, however, doesn't, so this has worked for me in testing so far:

```ini
[Service]
Type=oneshot
ExecStartPre=-modprobe zram
ExecStart=sh -c 'echo 1 > /sys/block/%i/reset'
ExecStart=zramctl /dev/%i --size 199229440
ExecStart=mkswap /dev/%i -L '%i'
```

Which yields the following journal
```
[ 7536.616068] nabtop systemd[1]: Deactivating swap Compressed swap on /dev/zram4...
[ 7536.637736] nabtop systemd[803]: dev-zram4.swap: Succeeded.
[ 7536.639312] nabtop systemd[1]: dev-zram4.swap: Succeeded.
[ 7536.640556] nabtop systemd[1]: Deactivated swap Compressed swap on /dev/zram4.
[ 7536.644219] nabtop systemd[1]: Starting Create swap on /dev/zram4...
[ 7536.729229] nabtop kernel: zram4: detected capacity change from 99229696 to 0
[ 7536.736443] nabtop kernel: zram4: detected capacity change from 0 to 199229440
[ 7536.746263] nabtop mkswap[24690]: Setting up swapspace version 1, size = 190 MiB (199225344 bytes)
[ 7536.746263] nabtop mkswap[24690]: LABEL=zram4, UUID=eb31760e-7711-4cf8-93f1-81b6a0862b1a
[ 7536.747291] nabtop systemd[1]: swap-create@zram4.service: Succeeded.
[ 7536.747864] nabtop systemd[1]: Started Create swap on /dev/zram4.
[ 7536.752237] nabtop systemd[1]: Activating swap Compressed swap on /dev/zram4...
[ 7536.770316] nabtop systemd[1]: Activated swap Compressed swap on /dev/zram4.
[ 7536.769215] nabtop kernel: Adding 194556k swap on /dev/zram4.  Priority:100 extents:1 across:194556k SSFS
```

**Comment by cmurf** [comment]
@nabijaczleweli I think the proposed change only masks the problem of swap-create@zram0.service running multiple times. What if it runs right after swapon? It would clobber an active swap device.

I think it's necessary to figure out why swap-create@zram0.service is running multiple times, and solve that problem.

**Comment by nabijaczleweli** [comment]
Well, hmm, the generator [will](https://www.freedesktop.org/software/systemd/man/systemd.generator.html#Description) be run multiple times, but if `swap-create@zram0.service` is running more than once then something other than us must be triggering it, since the the only activation route we install is `swap.target` -> `dev-zramX.swap` -> `swap-create@zramX.service`.

FWIW this is what I get on a fresh VM with just Rust and this installed, with a swap file at `/swap` and boots before installation removed:

```
nab@debst:~$ sudo journalctl -o short-monotonic | grep swap
[    0.412012] debst kernel: Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    2.546176] debst kernel: zswap: loaded using pool lzo/zbud
[   33.693708] debst systemd[1]: Created slice system-swap\x2dcreate.slice.
[   33.771664] debst systemd[1]: Activating swap /swap...
[   33.868235] debst kernel: Adding 1048572k swap on /swap.  Priority:-2 extents:2 across:1351676k FS
[   33.976353] debst mkswap[473]: Setting up swapspace version 1, size = 397 MiB (416280576 bytes)
[   33.976353] debst mkswap[473]: LABEL=zram0, UUID=ac298e4e-d6e7-419f-aa36-899ac1e4fdc0
[   33.979461] debst systemd[1]: Activated swap /swap.
[   34.029422] debst kernel: Adding 406524k swap on /dev/zram0.  Priority:100 extents:1 across:406524k SSFS
[   33.988545] debst systemd[1]: Starting Create swap on /dev/zram0...
[   33.989289] debst systemd[1]: swap-create@zram0.service: Succeeded.
[   33.989455] debst systemd[1]: Started Create swap on /dev/zram0.
[   33.996232] debst systemd[1]: Activating swap Compressed swap on /dev/zram0...
[   33.996401] debst systemd[1]: Activated swap Compressed swap on /dev/zram0.
[   61.254693] debst sudo[707]:      nab : TTY=tty1 ; PWD=/home/nab ; USER=root ; COMMAND=/usr/bin/systemctl list-dependencies swap.target
[   65.308965] debst sudo[710]:      nab : TTY=tty1 ; PWD=/home/nab ; USER=root ; COMMAND=/usr/sbin/swapon
[  262.597183] debst systemd[1]: Removed slice system-swap\x2dcreate.slice.
[  262.842454] debst systemd[1]: Deactivating swap Compressed swap on /dev/zram0...
[  262.845341] debst systemd[1]: Deactivating swap /swap...
[  262.865751] debst systemd[1]: swap.swap: Succeeded.
[  262.866433] debst systemd[1]: Deactivated swap /swap.
[  262.875504] debst systemd[1]: dev-zram0.swap: Succeeded.
[  262.876413] debst systemd[1]: Deactivated swap Compressed swap on /dev/zram0.
[    0.424307] debst kernel: Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    2.577960] debst kernel: zswap: loaded using pool lzo/zbud
[  101.936912] debst systemd[1]: Created slice system-swap\x2dcreate.slice.
[  102.012101] debst systemd[1]: Activating swap /swap...
[  102.084381] debst kernel: Adding 1048572k swap on /swap.  Priority:-2 extents:2 across:1351676k FS
[  102.090462] debst systemd[1]: Activated swap /swap.
[  102.095725] debst systemd[1]: Starting Create swap on /dev/zram0...
[  102.162853] debst mkswap[466]: Setting up swapspace version 1, size = 397 MiB (416280576 bytes)
[  102.162853] debst mkswap[466]: LABEL=zram0, UUID=ffea6c18-b090-494a-846a-f84c703c1fe0
[  102.180096] debst systemd[1]: swap-create@zram0.service: Succeeded.
[  102.180727] debst systemd[1]: Started Create swap on /dev/zram0.
[  102.188137] debst systemd[1]: Activating swap Compressed swap on /dev/zram0...
[  102.256540] debst kernel: Adding 406524k swap on /dev/zram0.  Priority:100 extents:1 across:406524k SSFS
[  102.220165] debst systemd[1]: Activated swap Compressed swap on /dev/zram0.
[  641.066808] debst systemd[1]: Removed slice system-swap\x2dcreate.slice.
[  641.290529] debst systemd[1]: Deactivating swap Compressed swap on /dev/zram0...
[  641.291565] debst systemd[1]: Deactivating swap /swap...
[  641.325637] debst systemd[1]: swap.swap: Succeeded.
[  641.326334] debst systemd[1]: Deactivated swap /swap.
[  641.327978] debst systemd[1]: dev-zram0.swap: Succeeded.
[  641.328993] debst systemd[1]: Deactivated swap Compressed swap on /dev/zram0.
[    0.415921] debst kernel: Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    2.600280] debst kernel: zswap: loaded using pool lzo/zbud
[   10.673346] debst systemd[1]: Created slice system-swap\x2dcreate.slice.
[   10.751011] debst systemd[1]: Activating swap /swap...
[   10.824194] debst kernel: Adding 1048572k swap on /swap.  Priority:-2 extents:2 across:1351676k FS
[   10.828141] debst systemd[1]: Activated swap /swap.
[   10.854546] debst systemd[1]: Starting Create swap on /dev/zram0...
[   11.939027] debst mkswap[451]: Setting up swapspace version 1, size = 397 MiB (416280576 bytes)
[   11.939027] debst mkswap[451]: LABEL=zram0, UUID=5041a14a-6a74-4f7f-ac32-38a961e33ea8
[   11.941188] debst systemd[1]: swap-create@zram0.service: Succeeded.
[   11.941887] debst systemd[1]: Started Create swap on /dev/zram0.
[   11.943860] debst systemd[1]: Activating swap Compressed swap on /dev/zram0...
[   12.012848] debst kernel: Adding 406524k swap on /dev/zram0.  Priority:100 extents:1 across:406524k SSFS
[   11.965250] debst systemd[1]: Activated swap Compressed swap on /dev/zram0.
[   50.838030] debst systemd[1]: Removed slice system-swap\x2dcreate.slice.
[   51.333171] debst systemd[1]: Deactivating swap Compressed swap on /dev/zram0...
[   51.334433] debst systemd[1]: Deactivating swap /swap...
[   51.354125] debst systemd[1]: dev-zram0.swap: Succeeded.
[   51.354811] debst systemd[1]: Deactivated swap Compressed swap on /dev/zram0.
[   51.356562] debst systemd[1]: swap.swap: Succeeded.
[   51.357389] debst systemd[1]: Deactivated swap /swap.
[    0.411357] debst kernel: Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    2.551798] debst kernel: zswap: loaded using pool lzo/zbud
[    6.419051] debst systemd[1]: Created slice system-swap\x2dcreate.slice.
[    6.512752] debst systemd[1]: Activating swap /swap...
[    6.599694] debst kernel: Adding 1048572k swap on /swap.  Priority:-2 extents:2 across:1351676k FS
[    6.571122] debst systemd[1]: Activated swap /swap.
[    6.636893] debst systemd[1]: Starting Create swap on /dev/zram0...
[    7.924902] debst mkswap[454]: Setting up swapspace version 1, size = 397 MiB (416280576 bytes)
[    7.924902] debst mkswap[454]: LABEL=zram0, UUID=7901ba75-16f8-4d4f-b7b5-b27ee02106ff
[    7.926910] debst systemd[1]: swap-create@zram0.service: Succeeded.
[    7.927541] debst systemd[1]: Started Create swap on /dev/zram0.
[    7.929417] debst systemd[1]: Activating swap Compressed swap on /dev/zram0...
[    8.232053] debst systemd[1]: Activated swap Compressed swap on /dev/zram0.
[    8.279657] debst kernel: Adding 406524k swap on /dev/zram0.  Priority:100 extents:1 across:406524k SSFS
```

It seems to be going up on boot with the other swaps and going down on shutdown likewise, as expected.

Manually starting `swap.target` with its dependencies (`swap.swap` and `dev-zram0.swap`) fulfilled does nothing, as expected, but manually starting either `swap-create@zram0.service` directly or via `dev-zram0.swap` yields an I/O error (presumably EBUSY, since the device is in use and so can't be reset).

In any case, resetting the device should be done anyway IMO, if only to protect against failures caused by other programs using it and not resetting if after they were done, it'll fail in a clear(er, I guess, if we did this in the generator we could handle EBUSY better, but an error on the reset line is better than on a set-size one) if the device is still in use, and will behave more as expected if it isn't.

**Comment by cmurf** [comment]
The service succeeds, therefore it should not ever run again. Upstream documentation says nothing about resetting the device before use. It's already in the reset state upon creation. Resetting it makes the problem worse by allowing swap-create@zram0.service to happen unlimited times. If swapon were to happen, and swap gets used, and later another swap-create@zram0.service happens with a device reset, that will result in a kernel panic.

That I so far have not been able to reproduce this with systemd.log_level=debug suggests there's a race or some other problem that's causing swap-create@zram0.service to be run more than once. It's supposed to be a one shot. Once it succeeds there's no reason it gets called again.

**Comment by nabijaczleweli** [comment]
Right; so far I haven't managed to reset a device that's in use but yeah, sure, this rationale makes sense to me.

Have you tried setting `RemainAfterExit=true` in the generated `swap-create@zramX.service`?

**Comment by cmurf** [comment]
I don't know how to do that. Since the service is generated, I think I'd have to modify the zram-generator?

**Comment by nabijaczleweli** [comment]
Not necessarily! You should be able to `systemctl edit swap-create@zram0.service` and write:
```ini
[Service]
RemainAfterExit=yes
```

Verify with `systemctl cat swap-create@zram0.service` – you should see the drop-in in `/etc` there.

**Comment by cmurf** [comment]
Following clean install of Fedora 32 following branch, I can't reproduce this in a VM after 10 attempts. I'll still give baremetal some more testing since it was fussier.

zram-generator-0.1.2-2.fc32.x86_64
systemd-245~rc1-3.fc32.x86_64
kernel-5.6.0-0.rc3.git0.1.fc32.x86_64

**Comment by cmurf** [comment]
I haven't seen this failure on VM or baremetal since the previous report. It seemed to go away with systemd 245.

**Comment by keszybz** [comment]
> ```ini
> [Service]
> RemainAfterExit=yes
> ```

Yeah, the service should have RemainAfterExit=yes set.

I'm still seeing this, and the proposed solution seems correct.

**Comment by keszybz** [comment]
Should be fixed by b2a7c8db000.
